homeassistant/components/nobo_hub/config_flow.py

```diff
@@ -0,0 +1,221 @@
+"""Config flow for Nobø Ecohub integration."""
+from __future__ import annotations
+
+import logging
+import socket
+from typing import Any
+
+from pynobo import nobo
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_COMMAND_OFF, CONF_COMMAND_ON, CONF_IP_ADDRESS
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import (
+    CONF_OVERRIDE_TYPE,
+    CONF_OVERRIDE_TYPE_CONSTANT,
+    CONF_OVERRIDE_TYPE_NOW,
+    CONF_SERIAL,
+    CONF_WEEK_PROFILE_NONE,
+    DOMAIN,
+)
+
+DATA_NOBO_HUB_IMPL = "nobo_hub_flow_implementation"
+DEVICE_INPUT = "device_input"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Nobø Ecohub."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.discovered_hubs = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if self.discovered_hubs is None:
+            self.discovered_hubs = await nobo.async_discover_hubs(loop=self.hass.loop)
+
+        errors = {}
+        if user_input is not None:
+            serial = user_input.get(CONF_SERIAL)
+            ip_address = user_input.get(CONF_IP_ADDRESS)
+            await self.async_set_unique_id(serial)
+            self._abort_if_unique_id_configured()
+            try:                
+                name = await self._test_connection(serial, ip_address)
+            except InvalidSerial:
+                errors["base"] = "invalid_serial"
+            except InvalidIP:
+                errors["base"] = "invalid_ip"
+            except DeviceNotFound:
+                errors["base"] = "device_not_found"
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            else:
+                return self.async_create_entry(title=name, data=user_input)
+
+
+        default_suggestion = self._prefill_identifier()
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_SERIAL, default=default_suggestion): str,
+                    vol.Optional(CONF_IP_ADDRESS): str,
```

Comment:
Is it serial or IP-based communication? I'm confused...

Reply:
It's TCP/IP. To connect, you have to provide the device's full serial number (12 digits). The IP address and the first 9 digits of the serial number may be discovered on the local network. Should I rename the configuration key?

Reply:
Ah, check 👍 

So why do we ask for it? As in, we can always ask for it in a second step?

Reply:
Serial number is always needed. IP address is optional, depending on your setup.

Reply:
> IP address is optional, depending on your setup.

Yes, that is what I meant. Can't we ask for IP if needed? (instead of always ask for it?)

Reply:
I guess the case were you have multiple hubs with the same serial number prefix on your local network can be solved by trying to connect to them in order and see which accepts. But what if we discover a hub on the local network, but you want to connect to a remote hub that's not auto discovered (e.g. over a VPN connection)? I don't think it's a problem to offer an optional field together with a required field:

![image](https://user-images.githubusercontent.com/5984518/148824003-89cc4e70-0d61-4124-9333-cb8f173fb6b7.png)

Otherwise there are two cases for requesting IP address:
1. No hub is found on the local network
2. HA is not able to connect to any hub on the local network

Should these two cases have different flows then? For the first HA should require an IP address right away, while for the latter HA should report a problem and request serial and IP address? This adds a bit of complexity to the config flow.

Reply:
I think making the common cases simpler from a user perspective is also worth the complexity when discovery finds devices to add.

Reply:
Implemented @allenporter 's suggestion.

Config window when HA discovers a hub on local network:
<img width="604" alt="image" src="https://user-images.githubusercontent.com/5984518/179365370-f023953d-65b7-48b4-91fb-a58b5225a625.png">


Config window if no hub is discovered, or the user checked "Configure undiscovered hub":
<img width="605" alt="image" src="https://user-images.githubusercontent.com/5984518/179365390-8280389d-90c2-4edd-93b6-25b4e945625b.png">